pypsa/linopf.py

```diff
@@ -257,6 +261,74 @@ def define_unit_commitment_status_variables(n, sns, c):
     define_binaries(n, (sns, com_i), c, "status", mask=active)
 
 
+def define_generator_multiple_device(n):
+    c = "Generator"
+    p_nom_i = n.generators.index[n.generators.p_nom > 0]
+    ext_i = get_extendable_i(n, c)
+    if (ext_i.intersection(p_nom_i)).empty:
+        return
+    ##serve? a mio parere no, quindi silenzio
+    # active = get_activity_mask(n, c, sns)[ext_i.intersection(p_nom_i)] if n._multi_invest else None
+    var = define_integers(n, (ext_i.intersection(p_nom_i)), "Generator", "n_opt")
+    define_boundary_for_multiple_device(n, c, "p_nom", np.array(var))
+    lhs = linexpr(
+        (
+            n.generators.p_nom[ext_i.intersection(p_nom_i)],
+            get_var(n, c, "n_opt")[ext_i.intersection(p_nom_i)],
+        ),
+        (-1, get_var(n, c, "p_nom")[ext_i.intersection(p_nom_i)]),
+    )
+    write_constraint(n, lhs, "=", 0)
+
+
+def define_boundary_for_multiple_device(n, c, attr, variables):
+    """
+    Initializes variables for nominal capacities for a given component and a
+    given attribute.
+
+    Parameters
+    ----------
+    n : pypsa.Network
+    c : str
+        network component of which the nominal capacity should be defined
+    attr : str
+        name of the variable, e.g. 'p_nom'
+    """
+    p_nom_i = n.generators.index[n.generators.p_nom > 0]
+    ext_i = get_extendable_i(n, c)
+    if (ext_i.intersection(p_nom_i)).empty:
+        return
+    if not (
+        (
+            n.df(c)[attr][ext_i.intersection(p_nom_i)]
+            < n.df(c)[attr + "_max"][ext_i.intersection(p_nom_i)]
+        )
+    ).all():
+        return
```

Comment:
If you are checking that p_nom is lower than p_nom_max, that is not necessary.
I wouldn't add such constraint: the code would work and should solve anyway, though this generator won't be installed